src/gui/dialogs/preferences_dialog.cpp

```diff
@@ -1213,9 +1212,14 @@ void preferences_dialog::handle_theme_select()
 
 void preferences_dialog::handle_gui2_theme_select()
 {
-	find_widget<label>(get_window(), "restart_msg", false).set_visible(widget::visibility::visible);
 	menu_button& gui2_theme_list = find_widget<menu_button>(this, "choose_gui2_theme", false);
-	prefs::get().set_gui_theme(gui2_themes_.at(gui2_theme_list.get_value()));
+	unsigned selected_theme = gui2_theme_list.get_value();
+	if (selected_theme != current_gui_theme_) {
+		current_gui_theme_ = selected_theme;
+		prefs::get().set_gui_theme(gui2_themes_.at(selected_theme));
+		gui2::init();
```

Comment:
Isn't this all that is needed? Why is the `RELOAD_GAME_DATA` stuff done? Mixing this with reloading game config does not sound good.

Reply:
`gui2::init()` by itself simply sets the current theme identifier and reads the theme, it does not change the appearance. (Think of it as roughly setting a string variable holding the current theme's name.)

The entire "hot reload" of this PR (immediately reloading the theme) comes from the `set_retval` line, which passes control to https://github.com/wesnoth/wesnoth/blob/72dc0390112437d4ba70eea07f5751b2be90050b/src/gui/dialogs/title_screen.cpp#L333-L337, which then reloads the UI. Without that you'll need to manually press F5  or restart to apply the newly selected theme.

Reply:
The second set_retval once again passes control to game_launcher IIRC, which then does the actual reloading.

Reply:
Well, `gui2::init()` also reads in all the GUI WML. (Something that shouldn't be needed here.) I suppose the other thing needed is actually redrawing the GUI.

Currently it looks like you don't just reload the theme. You reload the game config as well. I'd rather not mix the two.

Reply:
> Well, `gui2::init()` also reads in all the GUI WML. (Something that shouldn't be needed here.)

What about the case where someone installs a new theme from the add-on manager and wants to immediately switch to it? The GUI WML certainly needs to be reloaded in that case (though I'm not confident this is the correct place to do it).

Reply:
Found a way to stop loading the entire game_config. Apparently just reshowing the title_screen from the game_launcher scope was enough. This just closes then reshows the dialog so should not be much of an overhead.

Reply:
> What about the case where someone installs a new theme from the add-on manager and wants to immediately switch to it? The GUI WML certainly needs to be reloaded in that case (though I'm not confident this is the correct place to do it).

GUI WML should be reloaded as well with F5 or another key (but seems fine to reload "everything" with F5) and when reloading after addon install once addons can contain GUI WML. That doesn’t quite belong in this PR though.